### PR TITLE
(7zip*) migrated from ferventcoder

### PR DESCRIPTION
**Pull request by majkinetor**
- Package source: https://github.com/ferventcoder/chocolatey-packages/pull/223
- [3 open issues](https://github.com/ferventcoder/chocolatey-packages/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%207zip)

### CODE REVIEW

**Comment by pascalberger** [comment]
@majkinetor Will you change this also to an embedded package? Based on the [license](http://www.7-zip.org/license.txt) I think this should be possible.

**Comment by majkinetor** [comment]
> Will you change this also to an embedded package?

Yes, this should be embedded. You, rather then me, should do it, since you are migrating it now.  Its easier then creating online installer and people should get used to it. See [vlc](https://github.com/chocolatey/chocolatey-coreteampackages/tree/710f75060ffd8545ccbcdc24b0eea7b11ecadec2/automatic/vlc) as an example or any [other](https://github.com/chocolatey/chocolatey-coreteampackages/search?utf8=%E2%9C%93&q=get-remotefiles) that uses it.

In short:
- you can use `Get-RemoteFiles` in this repository (still unpublished) but this repo uses NEXT version.
- you will need to create embedded intaller which differes from online one (no checksums, urls etc.). I created `Get-RemoteFiles` in such way that changes to `chocolateyInstall` are trivial (in general case only silentArgs should be filled in and removed arch's that aren't used).

**Comment by majkinetor** [comment]
Ah, sorry, I thought I was writting on Git.

Yes, I will embedd this one.

**Comment by gep13** [comment]
@majkinetor `chocolatey` user has been added to each of these packages on chocolatey.org.

**Comment by ferventcoder** [comment]
This ready?

**Comment by AdmiringWorm** [comment]
@ferventcoder doesn't look like it, looks like it's currently just the same code as in your repo

**Comment by ferventcoder** [comment]
Right on.

**Comment by AdmiringWorm** [comment]
I believe this is done now, could someone please review it?

All issues at @ferventcoder is fixed except for https://github.com/ferventcoder/chocolatey-packages/issues/82 which I couldn't reproduce, making me think it have been fixed upstream

File: automatic/7zip.commandline/7zip.commandline.nuspec [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <metadata>
+    <id>7zip.commandline</id>
+    <title>7-Zip (Portable, CommandLine)</title>
+    <version>0.0</version>
+    <authors>Igor Pavlov</authors>
+    <owners>Rob Reynolds, chocolatey</owners>
```
Comment (pascalberger) [inline automatic/7zip.commandline/7zip.commandline.nuspec]: List `chocolatey` first as per [§1.2.4](https://github.com/chocolatey/chocolatey-coreteampackages/wiki/Contributing-guidelines#124-add-chocolatey-among-owners)

File: automatic/7zip.commandline/7zip.commandline.nuspec [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    <projectUrl>http://www.7-zip.org/</projectUrl>
+    <packageSourceUrl>https://github.com/chocolatey/chocolatey-coreteampackages/tree/master/automatic/7zip</packageSourceUrl>
+    <tags>7zip zip archiver foss</tags>
+    <licenseUrl>http://www.7-zip.org/license.txt</licenseUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <iconUrl>https://cdn.rawgit.com/chocolatey/chocolatey-coreteampackages/dbedb56d5ff709cd37f4abecc736d4a9e6400da8/icons/7zip.svg</iconUrl>
+    <docsUrl>http://www.7-zip.org/faq.html</docsUrl>
+    <mailingListUrl>https://sourceforge.net/p/sevenzip/discussion/45797/</mailingListUrl>
+    <bugTrackerUrl>https://sourceforge.net/p/sevenzip/_list/tickets?source=navbar</bugTrackerUrl>
+  </metadata>
```
Comment (pascalberger) [inline automatic/7zip.commandline/7zip.commandline.nuspec]: Add `<releaseNotes>` (http://www.7-zip.org/history.txt)

File: automatic/7zip.commandline/7zip.commandline.nuspec [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <metadata>
+    <id>7zip.commandline</id>
```
Comment (pascalberger) [inline automatic/7zip.commandline/7zip.commandline.nuspec]: Shouldn't we take the oppertunity and rename this to 7zip.portable?

File: automatic/7zip.commandline/tools/chocolateyInstall.ps1 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  }
+}
+Get-ChocolateyUnzip @packageArgs
+
+$packageArgs.packageName = '7zip.commandline Extras'
+$packageArgs.destination = "$toolsDir\7z-extra"
+$packageArgs.file        = $filePathExtra
+Get-ChocolateyUnzip @packageArgs
+
+Remove-Item -Path "$toolsDir\Uninstall.exe",$filePath32,$filePath64,$filePathExtra -Force
```
Comment (pascalberger) [inline automatic/7zip.commandline/tools/chocolateyInstall.ps1]: Add `-ea 0` to avoid install failing in case this fails, like it was done in other packages

File: automatic/7zip.commandline/update.ps1 [hunk 5]
```diff
@@ -0,0 +1,37 @@
+. "$PSScriptRoot\..\7zip\update.ps1"
+
+$softwareNamePrefix = '7-zip'
+
+function global:au_BeforeUpdate {
+  Get-RemoteFiles -Purge -FileNameBase '7zip'
+  $Latest.ChecksumType = 'sha256'
+
+  try {
+    $client = New-Object System.Net.WebClient
```
Comment (pascalberger) [inline automatic/7zip.commandline/update.ps1]: Should be outside of the try, since in case this throws an exception you don't want the finally block to be run, since it would throw another exception.

File: automatic/7zip.install/7zip.install.nuspec [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <metadata>
+    <id>7zip.install</id>
+    <title>7-Zip (Install)</title>
+    <version>0.0</version>
+    <authors>Igor Pavlov</authors>
+    <owners>Rob Reynolds, chocolatey</owners>
```
Comment (pascalberger) [inline automatic/7zip.install/7zip.install.nuspec]: List `chocolatey` first as per [§1.2.4](https://github.com/chocolatey/chocolatey-coreteampackages/wiki/Contributing-guidelines#124-add-chocolatey-among-owners)

File: automatic/7zip.install/7zip.install.nuspec [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    <licenseUrl>http://www.7-zip.org/license.txt</licenseUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <iconUrl>https://cdn.rawgit.com/chocolatey/chocolatey-coreteampackages/dbedb56d5ff709cd37f4abecc736d4a9e6400da8/icons/7zip.svg</iconUrl>
+    <docsUrl>http://www.7-zip.org/faq.html</docsUrl>
+    <mailingListUrl>https://sourceforge.net/p/sevenzip/discussion/45797/</mailingListUrl>
+    <bugTrackerUrl>https://sourceforge.net/p/sevenzip/_list/tickets?source=navbar</bugTrackerUrl>
+    <dependencies>
+      <dependency id="chocolatey-core.extension" version="1.0.4" />
+    </dependencies>
+  </metadata>
```
Comment (pascalberger) [inline automatic/7zip.install/7zip.install.nuspec]: Add `<releaseNotes>` (http://www.7-zip.org/history.txt)

File: automatic/7zip.install/tools/chocolateyInstall.ps1 [hunk 8]
```diff
@@ -0,0 +1,40 @@
+﻿Function Get-ExplorerProcessCount
```
Comment (pascalberger) [inline automatic/7zip.install/tools/chocolateyInstall.ps1]: Put into `helpers.ps1` as per [§.1.1.8](https://github.com/chocolatey/chocolatey-coreteampackages/wiki/Contributing-guidelines#118-clean-code)

File: automatic/7zip.install/tools/chocolateyInstall.ps1 [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+{
+  $process = Get-Process explorer -ErrorAction SilentlyContinue
+  $processCount = ($process | Measure-Object).Count
+  return $processCount
+}
+
```
Comment (pascalberger) [inline automatic/7zip.install/tools/chocolateyInstall.ps1]: Add `$ErrorActionPreference = 'Stop'`

File: automatic/7zip.install/tools/chocolateyInstall.ps1 [hunk 10]
```diff
@@ -0,0 +1,40 @@
+﻿Function Get-ExplorerProcessCount
+{
+  $process = Get-Process explorer -ErrorAction SilentlyContinue
```
Comment (pascalberger) [inline automatic/7zip.install/tools/chocolateyInstall.ps1]: `-ea 0` instead of `-ErrorAction SilentlyContinue` to be consistent with other packages

File: automatic/7zip.install/tools/chocolateyInstall.ps1 [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+Install-ChocolateyInstallPackage @packageArgs
+
+$finalProcessCount = Get-ExplorerProcessCount
+if($initialProcessCount -lt $finalProcessCount)
+{
+  Start-Process explorer.exe
+}
+
+Remove-Item "$filePath32*","$filePath64*" -Force -ErrorAction SilentlyContinue
```
Comment (pascalberger) [inline automatic/7zip.install/tools/chocolateyInstall.ps1]: `-ea 0` instead of `-ErrorAction SilentlyContinue` to be consistent with other packages

File: automatic/7zip.install/update.ps1 [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+function global:au_BeforeUpdate {
+  Get-RemoteFiles -Purge -FileNameBase '7zip'
+  $Latest.ChecksumType = 'sha256'
+}
+
+function global:au_SearchReplace {
+  @{
+    ".\tools\chocolateyInstall.ps1" = @{
+      "(?i)(^\s*softwareName\s*=\s*)'.*'" = "`$1'$softwareNamePrefix $($Latest.RemoteVersion)*'"
```
Comment (pascalberger) [inline automatic/7zip.install/update.ps1]: IMHO using `$Latest.RemoteVersion` is not 100% acurrate. It makes sense as long as the already installed version is the previous version, but what if I skip versions? But currently also don't have a better idea :(

File: automatic/7zip.install/update.ps1 [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    ".\tools\chocolateyUninstall.ps1" = @{
+      "(?i)(\s*\-SoftwareName\s+)'.*'" = "`$1'$softwareNamePrefix $($Latest.RemoteVersion)*'"
```
Comment (pascalberger) [inline automatic/7zip.install/update.ps1]: Shouldn't it be `$Latest.Version` here?

File: automatic/7zip/7zip.nuspec [hunk 14]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <metadata>
+    <id>7zip</id>
+    <title>7-Zip</title>
+    <version>0.0</version>
+    <authors>Igor Pavlov</authors>
+    <owners>Rob Reynolds, chocolatey</owners>
```
Comment (pascalberger) [inline automatic/7zip/7zip.nuspec]: List `chocolatey` first as per [§1.2.4](https://github.com/chocolatey/chocolatey-coreteampackages/wiki/Contributing-guidelines#124-add-chocolatey-among-owners)

File: automatic/7zip/7zip.nuspec [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    </description>
+    <projectUrl>http://www.7-zip.org/</projectUrl>
+    <packageSourceUrl>https://github.com/chocolatey/chocolatey-coreteampackages/tree/master/automatic/7zip</packageSourceUrl>
+    <tags>7zip zip archiver admin foss</tags>
+    <licenseUrl>http://www.7-zip.org/license.txt</licenseUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <iconUrl>https://cdn.rawgit.com/chocolatey/chocolatey-coreteampackages/dbedb56d5ff709cd37f4abecc736d4a9e6400da8/icons/7zip.svg</iconUrl>
+    <docsUrl>http://www.7-zip.org/faq.html</docsUrl>
+    <mailingListUrl>https://sourceforge.net/p/sevenzip/discussion/45797/</mailingListUrl>
+    <bugTrackerUrl>https://sourceforge.net/p/sevenzip/_list/tickets?source=navbar</bugTrackerUrl>
```
Comment (pascalberger) [inline automatic/7zip/7zip.nuspec]: Add `<releaseNotes>` (http://www.7-zip.org/history.txt)

File: automatic/7zip.commandline/7zip.commandline.nuspec (hunk 3 above)
Comment (AdmiringWorm) [inline automatic/7zip.commandline/7zip.commandline.nuspec]: sure, we can do that.
I just didn't have that as a priority at the moment

File: automatic/7zip.commandline/7zip.commandline.nuspec (hunk 1 above)
Comment (AdmiringWorm) [inline automatic/7zip.commandline/7zip.commandline.nuspec]: ah thanks, I forgot about that one.
Don't know why it's in there though, as the owners element isn't used on chocolatey.org as far as I know.

File: automatic/7zip.commandline/7zip.commandline.nuspec (hunk 2 above)
Comment (AdmiringWorm) [inline automatic/7zip.commandline/7zip.commandline.nuspec]: thanks, I wasn't able to find the changelog for 7zip

File: automatic/7zip.commandline/tools/chocolateyInstall.ps1 (hunk 4 above)
Comment (AdmiringWorm) [inline automatic/7zip.commandline/tools/chocolateyInstall.ps1]: will do

File: automatic/7zip.commandline/update.ps1 (hunk 5 above)
Comment (AdmiringWorm) [inline automatic/7zip.commandline/update.ps1]: this isn't .NET, so it don't matter if it's outside or inside the try/catch. (as far as I know)

File: automatic/7zip.install/tools/chocolateyInstall.ps1 (hunk 8 above)
Comment (AdmiringWorm) [inline automatic/7zip.install/tools/chocolateyInstall.ps1]: will do

File: automatic/7zip.install/tools/chocolateyInstall.ps1 (hunk 9 above)
Comment (AdmiringWorm) [inline automatic/7zip.install/tools/chocolateyInstall.ps1]: was sure I had it added, but will fix that one too.

File: automatic/7zip.install/update.ps1 (hunk 12 above)
Comment (AdmiringWorm) [inline automatic/7zip.install/update.ps1]: Actually, in this case it will be accurate.
The software name should be part of or equal to the name displayed in program and features

File: automatic/7zip.install/update.ps1 (hunk 13 above)
Comment (AdmiringWorm) [inline automatic/7zip.install/update.ps1]: no, $Latest.Version will be changed if there is a fix version, which will make the uninstall script unusable since it won't be able to find the uninstall key

File: automatic/7zip.commandline/update.ps1 (hunk 5 above)
Comment (AdmiringWorm) [inline automatic/7zip.commandline/update.ps1]: but I'll change it anyhow

**Comment by AdmiringWorm** [comment]
@pascalberger 
I believe I've added/responded to the feedback you provided.
With the exception of moving to 7zip.portable, will do that to but not yet.

**Comment by AdmiringWorm** [comment]
I believe the 7zip packages is now finally done.

/cc @pascalberger @ferventcoder

File: automatic/7zip/7zip.nuspec (hunk 15 above)
Comment (AdmiringWorm) [inline automatic/7zip/7zip.nuspec]: although it doesn't show up, I've added the release notes below this line

Comment (pascalberger) [inline automatic/7zip/7zip.nuspec]: @AdmiringWorm  Where and why doesn't it show up? Should be shown on chocolatey.org...

**Comment by pascalberger** [comment]
Thanks, LGTM

File: automatic/7zip/7zip.nuspec (hunk 15 above)
Comment (AdmiringWorm) [inline automatic/7zip/7zip.nuspec]: sorry, I meant it doesn't show where just commented as it was added on the line below

Comment (pascalberger) [inline automatic/7zip/7zip.nuspec]: Ah, comments in GitHub can be confusing :) I think I commented the closing metadata tag

Comment (AdmiringWorm) [inline automatic/7zip/7zip.nuspec]: you did for the other two packages.

**Comment by AdmiringWorm** [comment]
I'm just looking for a 👍 from @ferventcoder before this can be merged.

**Comment by AdmiringWorm** [comment]
I noticed the following packages need to be unrejected.
[7zip v16.04](https://chocolatey.org/packages/7zip/16.04)
[7zip.install v16.04](https://chocolatey.org/packages/7zip.install/16.04)

Not sure about [7zip.commandline v16.04](https://chocolatey.org/packages/7zip.commandline/16.04) needs to unrejected though, as we're deprecating it.

/cc @chocolatey/community-moderators

**Comment by gep13** [comment]
@AdmiringWorm I have unrejected the first two packages.  For the last one, I would be tempted to push out a package fix notation version, that way it is clear that it isn't an "official" release.

File: automatic/7zip.install/tools/chocolateyUninstall.ps1 [hunk 16]
```diff
@@ -0,0 +1,27 @@
+﻿$ErrorActionPreference = 'Stop';
+
+$packageName = '7zip.install'
+
+$uninstalled = $false
+[array]$key = Get-UninstallRegistryKey -SoftwareName '7-zip 16.04*'
```
Comment (ferventcoder) [inline automatic/7zip.install/tools/chocolateyUninstall.ps1]: Never a fan of passing the version. It doesn't upgrade well.

File: automatic/7zip.install/update.ps1 (hunk 12 above)
Comment (ferventcoder) [inline automatic/7zip.install/update.ps1]: Minus the version IMHO as well. If need be the major version.

File: automatic/7zip.install/update.ps1 (hunk 13 above)
Comment (ferventcoder) [inline automatic/7zip.install/update.ps1]: Why the version here?

File: deprecated/packages/7zip.commandline/7zip.commandline.nuspec [hunk 17]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <metadata>
+    <id>7zip.commandline</id>
+    <title>[Deprecated]7-Zip (Portable, CommandLine)</title>
```
Comment (ferventcoder) [inline deprecated/packages/7zip.commandline/7zip.commandline.nuspec]: Can we get a space after "]"? It should always be "[Deprecated] Title"

**Comment by ferventcoder** [comment]
Looks great. Just a few minor issues...

File: automatic/7zip.install/tools/chocolateyUninstall.ps1 (hunk 16 above)
Comment (AdmiringWorm) [inline automatic/7zip.install/tools/chocolateyUninstall.ps1]: Originally my reason for passing the version was to allow users to install multiple 7zip versions side-by-side, but noticed now that it doesn't matter since the registry key is overwritten anyhow.
I'll change it back to not pass the version.

Comment (ferventcoder) [inline automatic/7zip.install/tools/chocolateyUninstall.ps1]: Version 9.x and 15/16.x can be installed separately.

**Comment by AdmiringWorm** [comment]
@ferventcoder I've made the requested changes, if there isn't anything else I believe this can now be merged.

**Comment by ferventcoder** [comment]
LGTM

**Comment by AdmiringWorm** [comment]
👍 
The build is running to update the 7zip packages now, when it's done and the 7zip.portable package is approved I can push out the deprecated 7zip.commandline package
